arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/MemoizedDeepRecursiveFunction.kt

```diff
@@ -0,0 +1,38 @@
+package arrow.core
+
+import arrow.atomic.Atomic
+import arrow.atomic.loop
+
+/**
+ * Defines a recursive **pure** function that:
+ * - keeps its stack on the heap, which allows very deep recursive computations that do not use the actual call stack;
+ * - memoizes every call, which means that the function is execute only once per argument.
+ *
+ * [MemoizedDeepRecursiveFunction] takes one parameter of type [T] and returns a result of type [R].
+ * The [block] of code defines the body of a recursive function. In this block
+ * [callRecursive][DeepRecursiveScope.callRecursive] function can be used to make a recursive call
+ * to the declared function.
+ */
+public fun <T, R> MemoizedDeepRecursiveFunction(
+  block: suspend DeepRecursiveScope<T, R>.(T) -> R
+): DeepRecursiveFunction<T, R> {
+  val cache = Atomic(emptyMap<T, R>())
+  return DeepRecursiveFunction { x ->
+    when (x) {
+      in cache.get() -> cache.get().getValue(x)
```

Comment:
Wouldn't it be a little more performant to use:

```
cache.get().get(x)?.let {
   ...
}
```
?

Reply:
Yes, doing a single `get()` would probably be better 🤔 Not sure if this is correct syntax, but this could keep most of the code the same.

```
when(val map = cache.get()) {
  in map -> ...
  else -> ...
}
```